src/firefly/java/edu/caltech/ipac/firefly/server/query/ObsCorePackager.java

```diff
@@ -240,4 +288,137 @@ else if (obs_title != null) { //option 2, if productTitleTemplate is not availab
+                String raName = null, decName = null, cutoutName = null;
+
+                for (ParamInfo param : inputParams) {
+                    String ucd = param.getUCD();
+                    if (ucd == null) continue;
+
+                    if (isMatch(ucd, CUTOUT_UCDs)) cutoutName = param.getKeyName();
+                    else if (isMatch(ucd, RA_UCDs)) raName = param.getKeyName();
+                    else if (isMatch(ucd, DEC_UCDs)) decName = param.getKeyName();
+                }
+
+                if (raName != null && decName != null && cutoutName != null) {
+                    queryString.append(raName).append("=").append(ra).append("&")
+                            .append(decName).append("=").append(dec).append("&")
+                            .append(cutoutName).append("=").append(cutoutValue);
+                } else {
+                    return new CutoutServiceDescriptor("", new ArrayList<>()); // Invalid service descriptor
+                }
+            }
+
+            //Append other valid params (where value is not null or empty) (excluding RA, DEC, CUTOUT)
+            for (ParamInfo param : inputParams) {
+                String ucd = param.getUCD();
+                if (isMatch(ucd, CUTOUT_UCDs) || isMatch(ucd, RA_UCDs) || isMatch(ucd, DEC_UCDs)) continue;
+
+                String key = param.getKeyName();
+                String value = param.getStringValue();
+
+                if (key == null || key.isEmpty()) continue;
+
+                if (param.getRef() != null && (value == null || value.isEmpty())) {
+                    // Store missing params to be resolved later
+                    missingParams.add(new MissingParam(key, param.getRef()));
+                } else if (value != null && !value.isEmpty()) {
+                    queryString.append("&").append(key).append("=").append(value);
+                }
+            }
+        }
+
+        return new CutoutServiceDescriptor(accessUrl + "?" + queryString.toString(), missingParams);
+    }
+
+    //Utility function to check if UCD matches a known UCD category
+    private static boolean isMatch(String ucd, List<String> ucdArray) {
+        if (ucd == null || ucdArray == null || ucdArray.isEmpty()) {
+            return false;
+        }
+
+        String[] ucdParts = ucd.split(";");
+
+        //check if any ucsd exists in ucdArray
+        for (String part : ucdParts) {
+            if (ucdArray.contains(part.trim())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+
+    static class CutoutServiceDescriptor {
+        String partialUrl;
+        List<MissingParam> missingParams;
+
+        public CutoutServiceDescriptor(String partialUrl, List<MissingParam> missingParams) {
+            this.partialUrl = partialUrl;
+            this.missingParams = missingParams;
+        }
+    }
+
+    static class MissingParam {
+        String paramName;
+        String refId;
```

Comment:
This should be records. They are much nicer to work with. IntelliJ will do the conversion for you if you make all the object variables `final`. 
```java
record CutoutServiceDescriptor(String partialUrl, List<MissingParam> missingParams) { }

record MissingParam(String paramName, String refId) { }

```